adding  the Resource module doc (#21)
ansible-network/cli_rm_builder: docs/examples/rm_dev_guide.md

# THE RESOURCE MODULE DOC

### Is your model ready?

```

> git clone https://github.com/ansible-network/resource_module_models.git

Reviewer: This should be ansible-galaxy installation.

```

> Once the above repo is cloned go to the desired collection and add your model there. And talk to the team to get it _approved_.

Reviewer: ```suggestion
Once the above repo is cloned, go to the desired collection and add your model there. Reach out to the Ansible Network team to get it _approved_. You'll find us at #ansible-network on `irc.libera.chat <https://libera.chat/>_` or at `ansiblenetwork.slack.com`.
```

##### Helpful Links-

- [Ansible 101: Part 1: In the beginning there was YAML](https://www.redhat.com/en/blog/ansible-101-part-1-beginning-there-was-yaml)
- [YAML syntax](https://docs.ansible.com/ansible/latest/reference_appendices/YAMLSyntax.html)
- [Learn X in Y minutes](https://learnxinyminutes.com/docs/yaml/)

### Let’s get started with the Resource Module development.

At first, we would need a builder for our whole code base to get scaffolded from a tool - `cli_rm_builder`

```

> pip install ansible-base

Reviewer: ```suggestion
pip install ansible-core
```

ansible-galaxy collection install git+https://github.com/ansible-network/cli_rm_builder.git
```

The collection _dir_ should have the following structure, follow the Github _namespaces_ for the setup under ansible collection

```
~/../../collections
❯ tree -L 3
├── collections
│   └── ansible_collections
│       ├── ansible
│       │   ├── netcommon
│       │   └── utils

> │       ├── ansible_network

Reviewer: A RM dev should only use the galaxy installed cli_rm_builder. This might be a bit misleading since the sentence above has a _should_.

│       │   └── cli_rm_builder
│       ├── cisco
│       │   └── ios
│       ├── junipernetworks
│       │   └── junos
│       └── vyos
│           └── vyos

```

> Once the cli_rm_builder and the target collection repo is cloned in the desired location we can work on the development of our module.

Reviewer: ```suggestion
Once the cli_rm_builder is installed and the target collection repo is cloned in the desired location we can work on the development of our module.
```

```
❯ cat rm_builder_run.yml
---
- hosts: localhost
  gather_facts: yes
  roles:
    - ansible_network.cli_rm_builder.run
  vars:
    docstring:
    /../resource_module_models/models/{platform}/{{module}}/{platform}_{module_name}.yaml
    rm_dest:/home/{user}/{S}/{A}/collections/ansible_collections/{platform}/{platform}
    resource: {module_name}
    collection_org: {platform}
    collection_name: {platform}
    ansible_connection:local
```

```
❯ ansible-playbook rm_builder_run.yml
```

Post execution of the above command there should be few new files in your branch ready for the development of the resource module.

##### Important links at this point -

- Understanding RMs
  - [Ansible Network Resource Modules: Deep Dive on Return Values](https://www.ansible.com/blog/ansible-network-resource-modules-deep-dive-on-return-values)
  - [Developing network resource modules](https://docs.ansible.com/ansible/latest/network/dev_guide/developing_resource_modules_network.html)
  - [Getting started with Route Maps Resource Modules](https://www.ansible.com/blog/getting-started-with-route-maps-resource-modules)
    []()
- Debug Ansible resource module

>   - [Debugging Ansible Network Modules with VSCode](https://docs.google.com/document/d/1KtgzLG8N4cyQ35NunaxaM_Gl37gFd0LObTtq6kRlLK4/edit)

Reviewer: Is this doc public? If not, we should remove it and if possible make that information available publicly somewhere. Even a gist would work.

  - [Debugging modules](https://docs.ansible.com/ansible/latest/dev_guide/debugging.html)
- Setup development environment

>   - [Setting up Virtual Environments](https://docs.google.com/document/d/1eRUDIHZ6IYEegP9Z6HrhqWeGezM4n_8Pdxe8u-2TqN8/edit#heading=h.9qznvm65fex6)

Reviewer: Same as above.

>   - [Creation Of Virtual environment & Ansible Installation](https://github.com/rohitthakur2590/ansible-dev-virtualenv)

Reviewer: These might be out of scope for Resource Module development.

    []()

# Introduction to RM files -

> Looking at the collection where the new resource module is to be created after scaffolding the boiler plate code we should see new files already added -

Reviewer: ```suggestion
Looking at the collection where the new resource module is to be created after scaffolding the boiler plate code we should see the following new files already added -
```

> `{platform}_{module_name}.py` - the entry point to the Resource Module code and the module documentation resides here. To change or update any argspec attribute during development we need to make the required change here directly in the docstring and then rerun the rm_builder_run.yml with \_docstring var commented out.

Reviewer: ```suggestion
`{network_os}_{resource}.py` - the entry point to the Resource Module code and the module documentation resides here. To change or update any argspec attribute during development we need to make the required change here directly in the docstring and then rerun the rm_builder_run.yml with \_docstring var commented out.
```

```
../collections/ansible_collections/{platform}/{platform}/plugins/modules/{platform}_{module_name}.py
```

> `{module_name}.py` - under \_facts directory. The code in this file is responsible for converting native on-box configuration to Ansible structured data as per the argspec of the module by using the list of parsers defined in rm_templates/{module_name}.py.The on-box config can either be fetched from the target device or from the value set in the `running_config` key in the task when the module is run with `state: parsed`. The call to get data from the target device is often wrapped within a method. This is done to facilitate easier mocking when writing unit tests.

Reviewer: ```suggestion
`{resource}.py` - under /facts directory. The code in this file is responsible for converting native on-box configuration to Ansible structured data as per the argspec of the module by using the list of parsers defined in rm_templates/{module_name}.py.The on-box config can either be fetched from the target device or from the value set in the `running_config` key in the task when the module is run with `state: parsed`. The call to get data from the target device is often wrapped within a method. This is done to facilitate easier mocking when writing unit tests.
```

```

> ../collections/ansible_collections/{platform}/{platform}/plugins/module_utils/network/{platform}/facts/{module_name}/{module_name}.py

Reviewer: ```suggestion
../collections/ansible_collections/{network_os}/{network_os}/plugins/module_utils/network/{network_os}/facts/{resource}/{resource}.py
```

```

`facts.py` - You need to manually append the existence of your new resource module’s Facts class to the `FACTS_RESOURCE_SUBSET` dictionary in this file for facts to be generated as the call for facts is from a common instance i.e netcommon
The import -

```
from

> ansible_collections.{platform}.{platform}.plugins.module_utils.network.{platform}.facts.{module_name}.{module_name}

Reviewer: Same changes as above.

import (
  Logging_globalFacts,
)
```

The entry under global Var -

```
FACT_RESOURCE_SUBSETS = dict({module_name}={module_name}Facts,)
```

```

> ../collections/ansible_collections/{platform}/{platform}/plugins/module_utils/network/{platform}/facts/facts.p

Reviewer: Same as above.

y
```

> `logging_global.py` - The \_argspec file is the python level representation of your model. You may never need to edit it manually, change the model in the module file and cli_rm_builder should take care of updating it.

Reviewer: The "\_argspec" translates into "_argspec", this is incorrect. The folder name does not start with an underscore.

```

> ../collections/ansible_collections/{platform}/{platform}/plugins/module_utils/network/{platform}/argspec/{module_name}/{module_name}.py

Reviewer: Same as above.

```

> `logging_global.py` - the \_rm_template\* is one of the most vital components of a Resource Module since the conversion of native on-box configuration to structured data and vice-versa is facilitated by the parser templates that are defined in this file. Time to spin up regex/ jinja templating skills for this file. The better they are the easier it would be to get a higher score in the module's Unit Test Coverage (UTC).

Reviewer: Same as above.

```
../collections/ansible_collections/{platform}/{platform}/plugins/module_utils/network/{platform}/rm_templates/{module_name}.py
```

##### Helpful Links

- Regex parsers
  - [Pythex Regex parser](https://pythex.org/)
  - [Regex101](https://regex101.com/)
- Jinja2 parsers
  - [J2 Live Parser](http://jinja.quantprogramming.com/)
  - [Ansible Template Tester](https://ansible.sivel.net/test/)

> `Logging_global.py` - the \_config file contains all the core logic of how the execution should behave in various states. In here you get _want_ [the playbook] and _have_ [the config that came from the facts rendered]

Reviewer: Same as above.

All the states and their comparison logic goes here.

```
/home/sagpaul/Work/bannerNconfig/collections/ansible_collections/{platform}/{platform}/plugins/module_utils/network/{platform}/argspec/{module_name}/{module_name}.py
```

And, You might see a couple of *\_*init*\_*.py files generated,required for Ansible tests to pass!

## PHASE - 1 Gathering facts from the target device

The first step in building a resource module is to write facts code that converts device native configuration to structured data. This is done by comparing the device config with a set of pre-defined “Parser Templates” that define regexes to parse the native config.

Both the list of templates and the config are fed to an object of the NetworkTemplate class, on which the `parse()` method is then invoked. The output of the `parse()` method is semi-structured data that might need some additional updates to match the module’s argspec format.

> 

Reviewer: Can you add link a any module facts class here so that people reading can refer to it?

However, this comes in handy when the command to remove an attribute is significantly different from `setval`.

`compval`-
(Optional) This is to be used for a complex model where parsers are broken down into multiple ones and are then referenced like namespaces. By default, the name of the parser itself is used to extract an attribute for want and have dictionaries.

For example:

```
want = {'k1': {'k2': {'k3': 'newval', ‘k4’: 'anotherval'}}}
have = {'k1': {'k2': {'k3': 'oldval', ‘k4’: 'anotherval'}}}
```

With parser name `'k1.k2.k3'`, the RMEngineBase will extract the value of the nested key `'k3'` from both _want_ and _have_ and then compare it in order to decide if an update is required. In this case, it will compare `'k3': 'newval'` and `'k3': 'oldval'`. However, if the parser template has `compval: k1.k2` defined, the value of the key `k2`(which is a dictionary itself) will be used. So here, the comparison will happen between `'k2': {'k3':'newval', k4: 'anotherval'}` and `{'k3': 'oldval', k4: 'anotherval'}`

`shared`-
(Optional) The shared key makes the parsed values available to the rest of the parser

> entries until matched again.This enables the data/result of the parser to be shared among other parsers for reuse.

Reviewer: ```suggestion
entries until matched again. This enables the data/result of the parser to be shared among other parsers for reuse.
```

Example parsers -

```

> vyos@vyos:~$ show configuration commands | grep syslog

Reviewer: An IOS-styled example might be more relevant for a first time user of RMEngineBase.

set system syslog console facility all
set system syslog console facility local7 level 'err'
set system syslog console facility news level 'debug'
```

Given the set of commands the parser _can_ look like -
ref : [Vyos logging global model](https://github.com/ansible-network/resource_module_models/blob/master/models/vyos/logging_global/vyos_logging_global.yaml)

```
PARSERS =[{
            "name": "console.facilities",
            "getval": re.compile(
                r"""
                ^set\ssystem\ssyslog\sconsole\sfacility

>                 (\s(?P<facility>all|auth|authpriv|cron|daemon|kern|lpr|mail|mark|news|protocols|security|syslog|user|uucp|local[0-7]))?

Reviewer: We should not document running-config validation in parsers. This is redundant and hard to maintain in the long run.

                (\slevel\s(?P<level>'(emerg|alert|crit|err|warning|notice|info|debug|all)'))?
                $""", re.VERBOSE),
            "setval": tmplt_params,
            "remval": "system syslog console facility {{ console.facilities.facility }}",
            "result": {
                "console": {
                    "facilities": [{
                        "facility": "{{ facility }}",
                        "severity": "{{ level }}",
                    }, ]
                }
            }
        },]
```

> 

Reviewer: Please add a link to any `rm_templates/{{ resource }}.py` file for reference.

Having setval ready at this point is not required, we can start off by executing our first playbook

```
---
- name: check GATHERED state
  hosts: your.host.name
  gather_facts: no
  tasks:
  - name: Gather logging config
    vyos.vyos.vyos_logging_global:

>     state: gathered

Reviewer: ```suggestion
       state: gathered
```

```

```
---
- name: check PARSED state
  hosts: your.host.name
  gather_facts: no
  tasks:
  - name: Parse the provided configuration
    register: result
    vyos.vyos.vyos_logging_global:
      running_config: "{{ lookup('file', 'raw_vyos.cfg') }}"
      state: parsed
```

`raw_vyos.cfg`

This is just a flat file that holds the config that we get after the show commands are executed on the target device.

You should have a working facts code as of now! And the _gathered & parsed_ state should work before you proceed further.

##### Note

> If there is a _list of items_ in the generated facts, it is suggested to sort them before they are rendered, in order to to get consistent output across different Python versions. This also helps with assertions while working on Unit or IntegrationTests.

Reviewer: ```suggestion
If there is a _list of items_ in the generated facts, it is suggested to sort them before they are rendered, in order to to get consistent output across different Python versions. This also helps with assertions while working on Unit or Integration Tests.
```

> ## PHASE - 2 THE CONFIG / MERGED and other STATEs

Reviewer: This section is well documented in the below link and in various publicly available talks. Describing them once again with different definitions might confuse readers even more.

The set-based representations of the ACTION STATES are helpful and I'm +1 for keeping them.

Let’s talk about the different [states](https://docs.ansible.com/ansible/latest/network/user_guide/network_resource_modules.html#network-resource-module-states)!

`WANT` is basically the exact representation of the task, which is accessible in the config code.

`HAVE` is the on-box config as structured data that is rendered by the facts code output is just a representation of the commands that are formed based on want/have and states

`MERGED` -

```
|     WANT     |    HAVE     |    Output    |  Comment  |
| :----------: | :---------: | :----------: | :-------: |
| {A, B, C, D} |   {A,B,E}   |    {C,D}     |  Changed  |
|      {}      | {A,B,C,D,E} |      {}      | No change |
| {A, B, C, D} |     {}      | {A, B, C, D} |  Changed  |
```

`REPLACED`- _rA <- replace A_

```

|     WANT     |     HAVE     |     Output     |   Comment   |
| :----------: | :----------: | :------------: | :---------: |

> | {A, B, C, D} | {A, B, E, F} | {rA, rB, E, F} | Changed A,B |

Reviewer: This would also add 'C' and 'D' along side replacing 'A' and 'B'.

```

`OVERRIDDEN`- _nA - Negate A_

```
|     WANT     |    HAVE     |        Output         |  Comment  |
| :----------: | :---------: | :-------------------: | :-------: |

> | {A, B, C, D} |   {A,B,E}   |     {A, B, nE, D}     |  Changed  |

Reviewer: 'C' would also get added.

> |      {}      | {A,B,C,D,E} |          {}           | No change |

Reviewer: With overridden want cannot be empty.

Reviewer: If it is empty (i.e., check removed), it would behave like deleted.

| {A, B, C, D} |     {}      |     {A, B, C, D}      |  Changed  |
| {A, B, C, D} |   {E,F,G}   | {nE,nF,nG,A, B, C, D} |  Changed  |
```

> `DELETED`-

Reviewer: Please only document the delete all scenario.

Reviewer: Granular deletions are highly discouraged.

|      {}      | {A,B,C,D,E} | {nA,nB,nC,nD,nE} | No change |
| {A, B, C, D} |     {}      |        {}        | No Change |
| {A, B, C, D} |   {E,F,G}   |        {}        | No Change |
```

`RENDERED`- Pass in a config with the rendered state it is supposed to tell you all the set of commands that would be formed on the supplied config (without actually connecting to the target device), it is different from check mode.

`PARSED`- The parsed state is just opposite to the rendered state it tells you how the invocation/facts would look like when you supply the running_config/ raw config from a device.

##### Some important links at this point-

- Parsing semi-structured text with Ansible
  - [Parsing the CLI](https://docs.ansible.com/ansible/latest/network/user_guide/cli_parsing.html)
- Working with command output and prompts in network modules
  - [Handling prompts](https://docs.ansible.com/ansible/latest/network/user_guide/network_working_with_command_output.html)
- Network Debug and Troubleshooting Guide
  - [Debugging network Modules](https://docs.ansible.com/ansible/latest/network/user_guide/network_debug_troubleshooting.html)

> ## Conclusion

Reviewer: Shouldn't there be a separate section for config?
